// expensify-common/dist/CredentialsWrapper.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOGIN_PARTNER_DETAILS = void 0;
const localforage_1 = __importDefault(require("localforage"));
const LOGIN_PARTNER_DETAILS = {
    CREDENTIALS_KEY: 'DEVICE_SESSION_CREDENTIALS',
    EXPENSIFY_PARTNER_PREFIX: 'expensify.',
    PARTNER_NAME: 'chat-expensify-com',
    PARTNER_PASSWORD: 'e21965746fd75f82bb66',
};
exports.LOGIN_PARTNER_DETAILS = LOGIN_PARTNER_DETAILS;
const CredentialWrapper = {
    /**
     * @returns {Promise} Promise resolves to an object containing partnerUserID and partnerUserSecret
     */
    getCredentials() {
        return localforage_1.default.getItem(LOGIN_PARTNER_DETAILS.CREDENTIALS_KEY);
    },
    /**
     * @returns {Object} Object contains randomly generated partnerUserID and partnerUserSecret
     */
    generateCredentials() {
        return {
            partnerUserID: LOGIN_PARTNER_DETAILS.EXPENSIFY_PARTNER_PREFIX + crypto.randomUUID(),
            partnerUserSecret: crypto.randomUUID(),
        };
    },
    /**
     * @param {Object} credentials
     * @param {String} credentials.partnerUserID
     * @param {String} credentials.partnerUserSecret
     *
     * @returns {Promise} Promise resolves to an object containing partnerUserID and partnerUserSecret
     */
    setCredentials(credentials) {
        if (!credentials.partnerUserID || !credentials.partnerUserSecret) {
            // eslint-disable-next-line prefer-promise-reject-errors
            return Promise.reject('Invalid credential pair');
        }
        return localforage_1.default.setItem(LOGIN_PARTNER_DETAILS.CREDENTIALS_KEY, credentials);
    },
    /**
     * @returns {Promise}
     */
    clearCredentials() {
        return localforage_1.default.removeItem(LOGIN_PARTNER_DETAILS.CREDENTIALS_KEY);
    },
};
exports.default = CredentialWrapper;
