// expensify-common/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TLD_REGEX = exports.Str = exports.fastMerge = exports.Url = exports.Templates = exports.ReportHistoryStore = exports.PubSub = exports.PageEvent = exports.Num = exports.Network = exports.Logger = exports.ExpensiMark = exports.Device = exports.LOGIN_PARTNER_DETAILS = exports.CredentialsWrapper = exports.Cookie = exports.PUBLIC_DOMAINS_SET = exports.UI = exports.CONST = exports.g_cloudFrontImg = exports.g_cloudFront = exports.BrowserDetect = exports.APIDeferred = exports.API = void 0;
// eslint-disable-next-line rulesdir/no-api-in-views
var API_1 = require("./API");
Object.defineProperty(exports, "API", { enumerable: true, get: function () { return __importDefault(API_1).default; } });
var APIDeferred_1 = require("./APIDeferred");
Object.defineProperty(exports, "APIDeferred", { enumerable: true, get: function () { return __importDefault(APIDeferred_1).default; } });
var BrowserDetect_1 = require("./BrowserDetect");
Object.defineProperty(exports, "BrowserDetect", { enumerable: true, get: function () { return __importDefault(BrowserDetect_1).default; } });
var CONST_1 = require("./CONST");
Object.defineProperty(exports, "g_cloudFront", { enumerable: true, get: function () { return CONST_1.g_cloudFront; } });
Object.defineProperty(exports, "g_cloudFrontImg", { enumerable: true, get: function () { return CONST_1.g_cloudFrontImg; } });
Object.defineProperty(exports, "CONST", { enumerable: true, get: function () { return CONST_1.CONST; } });
Object.defineProperty(exports, "UI", { enumerable: true, get: function () { return CONST_1.UI; } });
Object.defineProperty(exports, "PUBLIC_DOMAINS_SET", { enumerable: true, get: function () { return CONST_1.PUBLIC_DOMAINS_SET; } });
var Cookie_1 = require("./Cookie");
Object.defineProperty(exports, "Cookie", { enumerable: true, get: function () { return __importDefault(Cookie_1).default; } });
var CredentialsWrapper_1 = require("./CredentialsWrapper");
Object.defineProperty(exports, "CredentialsWrapper", { enumerable: true, get: function () { return __importDefault(CredentialsWrapper_1).default; } });
Object.defineProperty(exports, "LOGIN_PARTNER_DETAILS", { enumerable: true, get: function () { return CredentialsWrapper_1.LOGIN_PARTNER_DETAILS; } });
exports.Device = __importStar(require("./Device"));
var ExpensiMark_1 = require("./ExpensiMark");
Object.defineProperty(exports, "ExpensiMark", { enumerable: true, get: function () { return __importDefault(ExpensiMark_1).default; } });
var Logger_1 = require("./Logger");
Object.defineProperty(exports, "Logger", { enumerable: true, get: function () { return __importDefault(Logger_1).default; } });
var Network_1 = require("./Network");
Object.defineProperty(exports, "Network", { enumerable: true, get: function () { return __importDefault(Network_1).default; } });
var Num_1 = require("./Num");
Object.defineProperty(exports, "Num", { enumerable: true, get: function () { return __importDefault(Num_1).default; } });
var PageEvent_1 = require("./PageEvent");
Object.defineProperty(exports, "PageEvent", { enumerable: true, get: function () { return __importDefault(PageEvent_1).default; } });
var PubSub_1 = require("./PubSub");
Object.defineProperty(exports, "PubSub", { enumerable: true, get: function () { return __importDefault(PubSub_1).default; } });
var ReportHistoryStore_1 = require("./ReportHistoryStore");
Object.defineProperty(exports, "ReportHistoryStore", { enumerable: true, get: function () { return __importDefault(ReportHistoryStore_1).default; } });
var Templates_1 = require("./Templates");
Object.defineProperty(exports, "Templates", { enumerable: true, get: function () { return __importDefault(Templates_1).default; } });
exports.Url = __importStar(require("./Url"));
var fastMerge_1 = require("./fastMerge");
Object.defineProperty(exports, "fastMerge", { enumerable: true, get: function () { return __importDefault(fastMerge_1).default; } });
var str_1 = require("./str");
Object.defineProperty(exports, "Str", { enumerable: true, get: function () { return __importDefault(str_1).default; } });
var tlds_1 = require("./tlds");
Object.defineProperty(exports, "TLD_REGEX", { enumerable: true, get: function () { return __importDefault(tlds_1).default; } });
