// expensify-common/dist/utils.js
"use strict";
'worklet';
Object.defineProperty(exports, "__esModule", { value: true });
exports.isWindowAvailable = isWindowAvailable;
exports.isNavigatorAvailable = isNavigatorAvailable;
exports.escapeText = escapeText;
exports.unescapeText = unescapeText;
exports.isFunction = isFunction;
exports.isObject = isObject;
/** Checks if the `window` global object is available. */
function isWindowAvailable() {
    return typeof globalThis.window !== 'undefined';
}
/** Checks if the `navigator` global object is available. */
function isNavigatorAvailable() {
    return typeof globalThis.navigator !== 'undefined';
}
const htmlEscapes = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&quot;',
    "'": '&#x27;',
    '`': '&#x60;',
};
const reUnescapedHtml = /[&<>"'`]/g;
const reHasUnescapedHtml = RegExp(reUnescapedHtml.source);
/**
 * Converts the characters "&", "<", ">", '"', and "'" in `string` to their
 * corresponding HTML entities.
 * Source: https://github.com/lodash/lodash/blob/main/src/escape.ts
 */
function escapeText(string) {
    return string && reHasUnescapedHtml.test(string) ? string.replace(reUnescapedHtml, (chr) => htmlEscapes[chr]) : string || '';
}
const htmlUnescapes = {
    '&amp;': '&',
    '&lt;': '<',
    '&gt;': '>',
    '&quot;': '"',
    '&#x27;': "'",
    '&#x60;': '`',
    '&#32;': ' ',
};
const reEscapedHtml = /&(?:amp|lt|gt|quot|#(x27|x60|32));/g;
const reHasEscapedHtml = RegExp(reEscapedHtml.source);
/**
 * The inverse of `escape`this method converts the HTML entities
 * `&amp;`, `&lt;`, `&gt;`, `&quot;` and `&#39;` in `string` to
 * their corresponding characters.
 * Source: https://github.com/lodash/lodash/blob/main/src/unescape.ts
 * */
function unescapeText(string) {
    return string && reHasEscapedHtml.test(string) ? string.replace(reEscapedHtml, (entity) => htmlUnescapes[entity] || "'") : string || '';
}
/**
 * Checks if the given variable is a function
 * @param {*} variableToCheck
 * @returns {boolean}
 */
function isFunction(variableToCheck) {
    return variableToCheck instanceof Function;
}
/**
 * Checks if the given variable is an object
 * @param {*} obj
 * @returns {boolean}
 */
function isObject(obj) {
    const type = typeof obj;
    return type === 'function' || (!!obj && type === 'object');
}
