diff --git a/node_modules/@10play/tentap-editor/src/RichText/Toolbar/actions.ts b/node_modules/@10play/tentap-editor/src/RichText/Toolbar/actions.ts
index d832667..4512f75 100644
--- a/node_modules/@10play/tentap-editor/src/RichText/Toolbar/actions.ts
+++ b/node_modules/@10play/tentap-editor/src/RichText/Toolbar/actions.ts
@@ -74,15 +74,6 @@ export const DEFAULT_TOOLBAR_ITEMS: ToolbarItem[] = [
       !editorState.isLinkActive && !editorState.canSetLink,
     image: () => Images.link,
   },
-  {
-    onPress:
-      ({ editor }) =>
-      () =>
-        editor.toggleTaskList(),
-    active: ({ editorState }) => editorState.isTaskListActive,
-    disabled: ({ editorState }) => !editorState.canToggleTaskList,
-    image: () => Images.checkList,
-  },
   {
     onPress:
       ({ setToolbarContext }) =>
@@ -146,30 +137,6 @@ export const DEFAULT_TOOLBAR_ITEMS: ToolbarItem[] = [
     disabled: ({ editorState }) => !editorState.canToggleBulletList,
     image: () => Images.bulletList,
   },
-  {
-    // Regular list items (li) and task list items both use the
-    // same sink command and button just with a different parameter, so we check both states here
-    onPress:
-      ({ editor, editorState }) =>
-      () =>
-        editorState.canSink ? editor.sink() : editor.sinkTaskListItem(),
-    active: () => false,
-    disabled: ({ editorState }) =>
-      !editorState.canSink && !editorState.canSinkTaskListItem,
-    image: () => Images.indent,
-  },
-  {
-    // Regular list items (li) and task list items both use the
-    // same lift command and button just with a different parameter, so we check both states here
-    onPress:
-      ({ editor, editorState }) =>
-      () =>
-        editorState.canLift ? editor.lift() : editor.liftTaskListItem(),
-    active: () => false,
-    disabled: ({ editorState }) =>
-      !editorState.canLift && !editorState.canLiftTaskListItem,
-    image: () => Images.outdent,
-  },
   {
     onPress:
       ({ editor }) =>
