// @expensify/react-native-live-markdown/lib/module/MarkdownTextInput.js
"use strict";

import { StyleSheet, TextInput, processColor } from 'react-native';
import React from 'react';
import { createWorkletRuntime, makeShareableCloneRecursive } from 'react-native-reanimated';
import MarkdownTextInputDecoratorViewNativeComponent from "./MarkdownTextInputDecoratorViewNativeComponent.js";
import NativeLiveMarkdownModule from "./NativeLiveMarkdownModule.js";
import { mergeMarkdownStyleWithDefault } from "./styleUtils.js";
import { jsx as _jsx } from "react/jsx-runtime";
let initialized = false;
let workletRuntime;
function getWorkletRuntime() {
  if (workletRuntime === undefined) {
    throw new Error("[react-native-live-markdown] Worklet runtime hasn't been created yet. Please avoid calling `getWorkletRuntime()` in top-level scope. Instead, call `getWorkletRuntime()` directly in `runOnRuntime` arguments list.");
  }
  return workletRuntime;
}
function initializeLiveMarkdownIfNeeded() {
  if (initialized) {
    return;
  }
  if (NativeLiveMarkdownModule) {
    NativeLiveMarkdownModule.install();
  }
  if (!global.jsi_setMarkdownRuntime) {
    throw new Error('[react-native-live-markdown] global.jsi_setMarkdownRuntime is not available');
  }
  workletRuntime = createWorkletRuntime('LiveMarkdownRuntime');
  global.jsi_setMarkdownRuntime(workletRuntime);
  initialized = true;
}
function registerParser(parser) {
  initializeLiveMarkdownIfNeeded();
  const shareableWorklet = makeShareableCloneRecursive(parser);
  const parserId = global.jsi_registerMarkdownWorklet(shareableWorklet);
  return parserId;
}
function unregisterParser(parserId) {
  global.jsi_unregisterMarkdownWorklet(parserId);
}
function processColorsInMarkdownStyle(input) {
  const output = JSON.parse(JSON.stringify(input));
  Object.keys(output).forEach(key => {
    const obj = output[key];
    Object.keys(obj).forEach(prop => {
      // TODO: use ReactNativeStyleAttributes from 'react-native/Libraries/Components/View/ReactNativeStyleAttributes'
      if (!(prop === 'color' || prop.endsWith('Color'))) {
        return;
      }
      obj[prop] = processColor(obj[prop]);
    });
  });
  return output;
}
function processMarkdownStyle(input) {
  return processColorsInMarkdownStyle(mergeMarkdownStyleWithDefault(input));
}
const MarkdownTextInput = /*#__PURE__*/React.forwardRef((props, ref) => {
  const markdownStyle = React.useMemo(() => processMarkdownStyle(props.markdownStyle), [props.markdownStyle]);
  if (props.parser === undefined) {
    throw new Error('[react-native-live-markdown] `parser` is undefined');
  }

  // eslint-disable-next-line no-underscore-dangle
  const workletHash = props.parser.__workletHash;
  if (workletHash === undefined) {
    throw new Error('[react-native-live-markdown] `parser` is not a worklet');
  }
  const parserId = React.useMemo(() => {
    return registerParser(props.parser);
  }, [props.parser]);
  React.useEffect(() => {
    return () => unregisterParser(parserId);
  }, [parserId]);
  return /*#__PURE__*/_jsx(MarkdownTextInputDecoratorViewNativeComponent, {
    style: styles.displayContents,
    markdownStyle: markdownStyle,
    parserId: parserId,
    children: /*#__PURE__*/_jsx(TextInput, {
      ...props,
      ref: ref
    })
  });
});
const styles = StyleSheet.create({
  displayContents: {
    display: 'contents'
  }
});
export default MarkdownTextInput;
export { getWorkletRuntime };
