// @expensify/react-native-live-markdown/lib/module/styleUtils.js
"use strict";

import { Platform } from 'react-native';
const FONT_FAMILY_MONOSPACE = Platform.select({
  ios: 'Courier',
  default: 'monospace'
});
const FONT_FAMILY_EMOJI = Platform.select({
  ios: 'System',
  android: 'Noto Color Emoji',
  default: 'System, Apple Color Emoji, Segoe UI Emoji, Noto Color Emoji'
});
function makeDefaultMarkdownStyle() {
  return {
    syntax: {
      color: 'gray'
    },
    link: {
      color: 'blue'
    },
    h1: {
      fontSize: 25
    },
    emoji: {
      fontSize: 20,
      fontFamily: FONT_FAMILY_EMOJI
    },
    blockquote: {
      borderColor: 'gray',
      borderWidth: 6,
      marginLeft: 6,
      paddingLeft: 6
    },
    code: {
      fontFamily: FONT_FAMILY_MONOSPACE,
      fontSize: 20,
      color: 'black',
      backgroundColor: 'lightgray',
      borderColor: 'gray',
      borderWidth: 1,
      borderRadius: 4,
      borderStyle: 'solid',
      padding: 0
    },
    pre: {
      fontFamily: FONT_FAMILY_MONOSPACE,
      fontSize: 20,
      color: 'black',
      backgroundColor: 'lightgray',
      borderColor: 'gray',
      borderWidth: 1,
      borderRadius: 4,
      borderStyle: 'solid',
      padding: 2
    },
    mentionHere: {
      color: 'green',
      backgroundColor: 'lime'
    },
    mentionUser: {
      color: 'blue',
      backgroundColor: 'cyan'
    },
    mentionReport: {
      color: 'red',
      backgroundColor: 'pink'
    },
    inlineImage: {
      minWidth: 50,
      minHeight: 50,
      maxWidth: 150,
      maxHeight: 150,
      marginTop: 5,
      marginBottom: 0,
      borderRadius: 5
    },
    loadingIndicator: {
      primaryColor: 'gray',
      secondaryColor: 'lightgray'
    }
  };
}
function mergeMarkdownStyleWithDefault(input) {
  const output = makeDefaultMarkdownStyle();
  if (input !== undefined) {
    Object.keys(input).forEach(key => {
      if (!(key in output)) {
        return;
      }
      const outputValue = output[key];
      if (outputValue) {
        Object.assign(outputValue, input[key]);
      }
    });
  }
  return output;
}
function parseStringWithUnitToNumber(value) {
  if (typeof value === 'number') {
    return value;
  }
  return value ? parseInt(value.replace('px', ''), 10) : 0;
}
export { mergeMarkdownStyleWithDefault, parseStringWithUnitToNumber };
