// react-native-marked/dist/module/components/MDTable.js
"use strict";

import React, { memo } from "react";
import { ScrollView, View } from "react-native";
import { Cell, Table, TableWrapper } from "react-native-reanimated-table";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
const MDTable = ({
  header,
  rows,
  widthArr,
  cellStyle,
  rowStyle,
  tableStyle,
  borderColor,
  borderWidth
}) => {
  return /*#__PURE__*/_jsx(ScrollView, {
    horizontal: true,
    children: /*#__PURE__*/_jsxs(Table, {
      borderStyle: {
        borderColor,
        borderWidth
      },
      style: tableStyle,
      children: [/*#__PURE__*/_jsx(TableWrapper, {
        style: rowStyle,
        children: header.map((headerCol, index) => {
          return /*#__PURE__*/_jsx(Cell, {
            width: widthArr[index],
            data: /*#__PURE__*/_jsx(View, {
              style: cellStyle,
              children: headerCol
            })
          }, `${index}`);
        })
      }), rows.map((rowData, index) => {
        return /*#__PURE__*/_jsx(TableWrapper, {
          style: rowStyle,
          children: rowData.map((cellData, cellIndex) => {
            return /*#__PURE__*/_jsx(Cell, {
              width: widthArr[cellIndex],
              data: /*#__PURE__*/_jsx(View, {
                style: cellStyle,
                children: cellData
              })
            }, `${cellIndex}`);
          })
        }, `${index}`);
      })]
    })
  });
};
export default /*#__PURE__*/memo(MDTable);
